src/Templates/NServiceBusWindowsService/Host.cs

```diff
@@ -0,0 +1,82 @@
+
+namespace NServiceBusWindowsService
+{
+    class Host
+    {
+        // TODO: optionally choose a custom logging library
+        // https://docs.particular.net/nservicebus/logging/#custom-logging
+        // LogManager.Use<TheLoggingFactory>();
+        static readonly ILog log = LogManager.GetLogger<Host>();
+
+        IEndpointInstance endpoint;
+
+        // TODO: give the endpoint an appropriate name
+        public string EndpointName => "MyNServiceBusWindowsService";
+
+        public async Task Start()
+        {
+            try
+            {
+                var endpointConfiguration = new EndpointConfiguration(EndpointName);
+                endpointConfiguration.DefineCriticalErrorAction(OnCriticalError);
+
+                endpointConfiguration.ApplySettingsForAllEndpoints(transport =>
+                {
+                    // TODO: apply endpoint-specific transport settings
+                });
+
+                // TODO: perform any futher start up operations before or after starting the endpoint
+                endpoint = await Endpoint.Start(endpointConfiguration);
+            }
+            catch (Exception ex)
+            {
+                await LogFatalAndFailFast("Failed to start.", ex);
+            }
+        }
+
+        public async Task Stop()
+        {
+            try
+            {
+                // TODO: perform any futher shutdown operations before or after stopping the endpoint
+                await endpoint?.Stop();
+            }
+            catch (Exception ex)
+            {
+                await LogFatalAndFailFast("Failed to stop correctly.", ex);
+            }
+        }
+
+        async Task OnCriticalError(ICriticalErrorContext context)
+        {
+            // TODO: decide if stopping the endpoint and exiting the process is the best response to a critical error
+            // https://docs.particular.net/nservicebus/hosting/critical-errors
+            // and consider setting up service recovery
+            // https://docs.particular.net/nservicebus/hosting/windows-service?version=core_7#installation-restart-recovery
+            try
+            {
+                await context.Stop();
+            }
+            finally
+            {
+                await LogFatalAndFailFast($"Critical error, shutting down: {context.Error}", context.Exception);
+            }
+        }
+
+        async Task LogFatalAndFailFast(string message, Exception exception)
+        {
+            log.Fatal(message, exception);
+
+            // TODO: when using an external logging framework it is important to flush any pending entries prior to calling FailFast
+            // https://docs.particular.net/nservicebus/hosting/critical-errors#when-to-override-the-default-critical-error-action
+            await Task.CompletedTask;
+
+            // TODO: https://docs.particular.net/nservicebus/hosting/windows-service#installation-restart-recovery
+            Environment.FailFast(message, exception);
```

Comment:
I previously had this in a `finally`, with the rest of the method in a `try`, but because this is a `FailFast()`, it would mask any exceptions thrown in the `try`, effectively causing them to be swallowed. However, that means that if there is an exception in the log flushing, that error will bubble out of the process instead of the exception that caused the problem in the endpoint, and if we have to choose between the two, we probably want the exception from the endpoint rather than the exception from the log flushing, so perhaps I should re-instate the `try...finally`?

Reply:
What exceptions would be swallowed?  What else would be in this method that you're concerned about? If the logger throws when we're shutting down anyway, that seems fine to me.

Reply:
The problem is, that if the logging framework throws then that exception will be thrown out to the runtime as the reason that the process stopped, and the exception that actually caused the endpoint to shutdown in the first place, while still being a first chance exception and therefore perhaps being logged somewhere,  is effectively being swallowed. I think it may be better for me to re-instate the `try..finally`, so that the log fatal/log flushing becomes "best effort" and we ensure that the exception being passed to the method is the one that the runtime sees as causing the process to stop via the fail-fast.